### PR TITLE
make glVertexAttribIPointer work better with client-side vertex arrays

### PR DESCRIPTION
**Pull request by hujiajie**
In WebGL 2.0, `drawArrays` generates an `INVALID_OPERATION` error if `vertexAttribPointer` is used and the base type of the corresponding shader attribute is not floating-point, so `glVertexAttribPointer` and `glVertexAttribIPointer` should be distinguished while emulating client-side vertex arrays in FULL_ES3 mode.

### CODE REVIEW

**Comment by hujiajie** [comment]
cc @gyagp

File: src/library_gl.js [hunk 1]
```diff
@@ -350,7 +350,7 @@ var LibraryGL = {
 #if GL_ASSERTIONS
         GL.validateVertexAttribPointer(cb.size, cb.type, cb.stride, 0);
 #endif
-        GLctx.vertexAttribPointer(i, cb.size, cb.type, cb.normalized, cb.stride, 0);
+        cb.vertexAttribPointer.call(GLctx, i, cb.size, cb.type, cb.normalized, cb.stride, 0);
```
Comment (kripken) [inline src/library_gl.js]: Perhaps we should rename the function here. As it may refer to either `vertexAttribPointer` or `vertexAttribIPointer`, a more generic name that implies it could be either might be better. Perhaps something generic like `setup`, or maybe specific like `vertexAttribXPointer` (X implying it might be one of multiple values)? A benefit of a more unique name is that a search for that name in this file would find just the specific things that this property is used for.

File: src/library_gl.js [hunk 2]
```diff
@@ -7611,6 +7611,7 @@ var LibraryGL = {
       cb.stride = stride;
       cb.ptr = ptr;
       cb.clientside = true;
+      cb.vertexAttribPointer = GLctx.vertexAttribPointer;
```
Comment (kripken) [inline src/library_gl.js]: this is actually risky. I'm not sure what happens if the context is destroyed or replaced, but I strongly suspect a reference to a function on it would be invalid. Instead, creating a function like in the case lower down is guaranteed to be safe.

File: tests/test_browser.py [hunk 3]
```diff
@@ -1681,6 +1681,11 @@ def test_gles2_emulation(self):
                        '-s', 'FULL_ES2=1', '-lGL', '-lEGL', '-lX11',
                        '--preload-file', 'basemap.tga', '--preload-file', 'lightmap.tga', '--preload-file', 'smoke.tga'])
 
+  @requires_graphics_hardware
+  def test_clientside_vertex_arrays_es3(self):
+    # NOTE: Should FULL_ES3=1 imply client-side vertex arrays? The emulation needs FULL_ES2=1 for now.
```
Comment (kripken) [inline tests/test_browser.py]: I agree with this comment - we should turn on `FULL_ES2` when `FULL_ES3` is set. Can do that in emcc.py. This could be left to another PR though, but if so we should file an issue so we don't forget it, unless you think you can get to in very soon.

File: src/library_gl.js (hunk 2 above)
Comment (hujiajie) [reply src/library_gl.js]: First I'd like to point out that `GLctx.vertexAttribPointer` is `GLctx.__proto__.vertexAttribPointer`, which is either `WebGLRenderingContext.prototype.vertexAttribPointer` or `WebGL2RenderingContext.prototype.vertexAttribPointer`, so `cb.vertexAttribPointer` always refers to the same function object.
Then what if `GLctx` is replaced or destroyed? Maybe it would result in *undefined behavior*. A sane application should be responsible for setting up the vertex arrays properly when it creates a new context, and draw functions should never be called after the context is destroyed.
I still made the requested change though because it seems to be more eye-candy.

File: tests/test_browser.py (hunk 3 above)
Comment (hujiajie) [reply tests/test_browser.py]: In fact I do not mean to turn on `FULL_ES2` when `FULL_ES3` is set (I'm neutral to this), and the documentation says [these two flags are orthogonal](http://kripken.github.io/emscripten-site/docs/porting/multimedia_and_graphics/OpenGL-support.html#opengl-es-2-0-3-0-emulation). I'm just wondering if client-side vertex arrays should be enabled here. It's a feature supported by ES2, but discouraged in ES3.

Comment (kripken) [reply tests/test_browser.py]: Thanks, good point, I didn't realize that about those flags.

Comment (kripken) [reply tests/test_browser.py]: (Not sure about the answer to the question about client-side arrays.)

**Comment by kripken** [comment]
Thanks @hujiajie, the changes here look good, merging.
